Support encrypted packets

tristanbatchler opened the pull request:
See https://moonlapsemud.atlassian.net/browse/MOON-11 for feature description.

Implementation notes:
1. Introduce a new class (implementations on both client- and server-side) called `PacketConfig`
  * This class allows easy setting of eight flags, most of which are reserved for now. The two which are important are called `rsa_encrypted` and `aes_encrypted`.
  * This class has a `to_bytes` and `from_bytes` method, allowing to switch back and forth between a 1-byte header and the human-readable `PacketConfig` object.
  * The byte generated with the `PacketConfig.to_bytes` method is prepended to each packet during sending, and is read off easily upon receive.

2. Introduce `encrypted` flag on protobuf packet types
  * Each type under the base `Packet` definition in `packets.proto` now has an optional `[(encrypted) = true]` option, which marks the sub-packet as requiring encryption. You can explicitly define `[(encrypted) = false]` too, if you want, but this is not required.

3. Introduce a `CryptoContext` class (implementations on both client- and server-side), which provides implementations for all required cryptographic algorithms (RSA and AES), and also provides a means to keep track of keys
  * Each client and each protocol will keep its own `CryptoContext` object, which keeps track of the keys required, and allows easy encryption/decryption using those keys on packets that require them.

tristanbatchler commented:
Notes: 
1. `PacketConfig.HasFlag` function is O(n) each time it's called (each time we send a packet!). We should make a dictionary in memory at startup for easy O(1) lookup on each send instead.
2. Currently, what's stopping a packet from having "RSAEncrypted" AND "AESEncrypted" flags set? What would happen?
3. Is it really a good idea to allow the client to generate their own AES keys?


Side note:
1. Sending over socket doesn't support flushing (even C# Flush method does nothing). Need to have a separator or prepended length to avoid the receiving end thinking consecutive packets are all part of the same data stream.

trithagoras commented on Server/Packets/CryptoContext.cs:
```diff
@@ -0,0 +1,107 @@
+        // The RSA keys should go in the application base directory
+        readonly static string keysPath = Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "Keys");
+
+        private byte[]? clientAESPrivateKey;
```
```suggestion
        byte[]? clientAESPrivateKey;
```

`private` is redundant

trithagoras commented on Server/Packets/CryptoContext.cs:
```diff
@@ -0,0 +1,107 @@
+        readonly static string keysPath = Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "Keys");
+
+        private byte[]? clientAESPrivateKey;
+        private RSACryptoServiceProvider serverRSA;
```
```suggestion
        RSACryptoServiceProvider serverRSA;
```

trithagoras commented on Server/Packets/PacketDeliveryService.cs:
```diff
@@ -15,27 +15,53 @@ public class PacketDeliveryService : IPacketDeliveryService {
-            data = data[0..bytesRead];  // strip trailing empty bytes
+            if (packetConfig.RSAEncrypted) {
+                data = cryptoContext.RSADecrypt(data);
+            }
```
no `else` statement here allows a packet to be both AES and RSA encrypted

tristanbatchler commented:
@trithagoras I have fixed the following issues and done extensive testing with the debugger to make sure it's working correctly.

> *PacketConfig.HasFlag function is O(n) each time it's called (each time we send a packet!). We should make a dictionary in memory at startup for easy O(1) lookup on each send instead.*

I included a static, persistent cache dictionary which looks something like this once populated:
```
{
        "LoginPacket": { "Encrypted": true },
        "ChatPacket": { "Encrypted": false }
}
```
We don't fully populate this cache at the start as initially discussed; rather we populate it on-the-fly so the `HasFlag` algorithm is O(n) the first time the server encounters a packet and flag combo, but O(1) every subsequent time. I chose to do this so we're not relying on the server/client to call some `PopulateCache` function during startup. This technique is called [memoization](https://en.wikipedia.org/wiki/Memoization).

> *Currently, what's stopping a packet from having "RSAEncrypted" AND "AESEncrypted" flags set? What would happen?*

I implemented your suggested fix (adding an `else if` statement instead of another `if` statement in `Server/PacketDeliveryService.cs` line 39. The client side already has an `elif` statement in its equivalent area (`Client/tcp.py` line 51).

> *Is it really a good idea to allow the client to generate their own AES keys?*

We need to research this. Pros & cons:
**Pros:**
1. The server doesn't have to generate an AES key for each client connecting -> easier on server CPU time

**Cons:**
1. If the client code is altered to generate insecure AES keys, their own encryption could be compromised. 

The con here *sounds* bad, but if somebody did this to their client, they should understand the risk they are putting on **themselves**. It's simply impossible for these guys to affect the security of other clients, though, unless they are distributing modified copies of the client and snooping on packets sent to/from their victims. There are two ways to address this issue and I think we should choose the least expensive option:
1. The server validates the client's AES key when it receives it. I don't know how this would be done though, and I don't even know if it is possible.
2. The server generates AES keys for the client. Then the question is "how do we get it to the client securely"? It would require the client to have an RSA keypair and then we're back to square one with "is it a good idea to let the client generate their own keys"?

There must be something we're missing here because, for example, Git lets each client generate their own keys. We need to think about this.

> *private is redundant*

I've fixed both those redundant `private` keywords.

trithagoras commented on Server/Packets/CryptoContext.cs:
```diff
@@ -0,0 +1,107 @@
+        public CryptoContext() {
+            serverRSA = new RSACryptoServiceProvider(2048);
+
+            string publicKey;
+            string privateKey;
+            try {
+                publicKey = File.ReadAllText(Path.Join(keysPath, "public.pem"));
+                privateKey = File.ReadAllText(Path.Join(keysPath, "private.pem"));
+            }
+            catch (IOException e) {
+                Log.Error("Error while reading keys: " + e.Message);
+                Log.Error("Did you generate the RSA KeyPair first (use CryptoContext.GenerateRSAKeyPair())?");
+                throw e;
+            }
```
So because this is in the constructor, file reading will be done for each protocol that connects. This is potentially a lot of duplicate memory use.

tristanbatchler replied:
Addressed in c2108f450842661c942ad9290e70890fad3746af

trithagoras commented on Server/Packets/CryptoContext.cs:
```diff
@@ -0,0 +1,107 @@
+using Serilog;
+
+namespace Moonlapse.Server.Packets {
+    public class CryptoContext {
```
It seems to be that this entire class could (maybe?) be a `Service` that we can inject into the `Container`, thus eliminating the need to have it in the signature of other methods and eliminating the need to have a `CryptoContext` object per `Protocol`.

If so, then we could make this implement an interface and register it in the startup and just use `Container.ResolveRequired<ICryptoContext>()` wherever it is required. We can then make `PacketDeliveryService` depend on an implementation of `ICryptoContext`.

trithagoras replied:
A downside to this is that public key and private key will never be refreshed unless the server is restarted. Something to consider in a future issue, perhaps.

tristanbatchler replied:
Addressed in aa4a49fc09b64d299232abbf2cd1de1cb36286ec (I renamed the CryptoContext service later to CryptoContextService  - 086ee7c96bc3b197a54aed6db6de321609a98b0c)

trithagoras commented on Server/Packets/IPacketDeliveryService.cs:
```diff
@@ -7,7 +7,7 @@
     public interface IPacketDeliveryService {
-        public Task SendPacketAsync(NetworkStream stream, Packet packet);
-        public Task<Packet> ReceivePacketAsync(NetworkStream stream);
+        public Task SendPacketAsync(NetworkStream stream, Packet packet, CryptoContext cryptoContext, PacketConfig? config = default);
+        public Task<Packet> ReceivePacketAsync(NetworkStream stream, CryptoContext cryptoContext);
```
As per the suggestions in `CryptoContext.cs`, I'd prefer if `PacketDeliveryService` relies on an implementation of `ICryptoContext` so that a concrete `CryptoContext` doesn't have to be specified in the method signatures. It'll just make it more flexible for possible future changes & testing, etc.

tristanbatchler replied:
Addressed in aa4a49fc09b64d299232abbf2cd1de1cb36286ec (I renamed the CryptoContext service later to CryptoContextService  - 086ee7c96bc3b197a54aed6db6de321609a98b0c)

trithagoras commented on Server/Packets/PacketConfig.cs:
```diff
@@ -0,0 +1,125 @@
+        /// }
+        /// </code>
+        /// </summary>
+        static Dictionary<string, Dictionary<Extension<FieldOptions, bool>, bool>> flagsCache = new();
```
global state is generally seen as less favourable than injecting singleton services. https://stackoverflow.com/questions/7026507/why-are-static-variables-considered-evil

It may be a good idea to separate what can be considered static into another service to be injected (or possibly could even be introduced into the `PacketDeliveryService`, but I think I'd prefer it to be another service that `PacketDeliveryService` depends on).

tristanbatchler replied:
Addressed in c44732c92d34bce79250ca3fa84b1763b7bc4c17

trithagoras commented on Server/Packets/PacketDeliveryService.cs:
```diff
@@ -15,27 +15,52 @@ public class PacketDeliveryService : IPacketDeliveryService {
+            config ??= new PacketConfig();
+
+            // Ensure the AESEncrypted flag is set on the header if the packet type demands encryption
+            if (PacketConfig.HasFlag(packet, PacketsExtensions.Encrypted)) {
+                config.AESEncrypted = true;
+                config.RSAEncrypted = false;
+            }
```
I think it's a little bit confusing that you are able to set the encryption flags when creating the `PacketConfig` object, but it immediately gets overridden when you try to attach it and send a packet. It might be clearer if it's not possible to set the encryption flags when creating the object, or to have some sort of possible override instead.

tristanbatchler replied:
No longer relevant after c44732c92d34bce79250ca3fa84b1763b7bc4c17 I believe

trithagoras commented on Server/Packets/PacketConfig.cs:
```diff
@@ -0,0 +1,125 @@
+            Reserved6 = reserved6;
+        }
+
+        public static PacketConfig FromByte(byte b) {
```
As in the previous comment, (_global state is generally seen as less favourable than injecting singleton services_), i think these static methods could be moved into the same service. Also, for stylish purposes, anything static should appear above the constructor and field declarations.

tristanbatchler replied:
Addressed in c44732c92d34bce79250ca3fa84b1763b7bc4c17 I believe

trithagoras commented on Client/tcp.py:
```diff
@@ -11,20 +16,68 @@ def __init__(self, hostname: str, port: int):
+                    aes_key_packet = pack.Packet()
+                    aes_key_packet.aes_key.key = self.crypto_context.get_client_aes_private_key()
+                    self.send_packet(aes_key_packet)
+                    time.sleep(1) # Required to ensure the server has received the AES key before we try to login
```
as discussed the other day, the need for this will be fixed in MOON-13.

tristanbatchler replied:
OK
